[quant] Quantized comparators

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24387 [quant] Quantized comparators**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24387

Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Commit:
Quantized comparators

zafartahirov commented:
Please, note -- this is the version that uses `dequantize`. If we need to implement the optimized one -- I can write a tensoriterator in the macro that would.

Commit:
Update on "Quantized comparators"


Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Commit:
Update on "Quantized comparators"


Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Commit:
Update on "Quantized comparators"


Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

jerryzh168 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3613,133 +3613,157 @@
   dispatch:
     CPU: ne_scalar_out_cpu
     CUDA: ne_scalar_out_cuda
+    QuantizedCPU: quantized_ne_out
```
as we discussed, name should be `ne_scalar_out_quantized_cpu` in this case

zafartahirov replied:
Even if can overload them?

jerryzh168 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3613,133 +3613,157 @@
   dispatch:
     CPU: legacy::cpu::_th_ne
     CUDA: legacy::cuda::_th_ne
+    QuantizedCPU: quantized_ne
```
similar for all the other functions

Commit:
Update on "Quantized comparators"


Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Tests:

```
$> python test/test_quantized.py --v -- TestComparatorOps
test_compare_tensor_scalar (__main__.TestComparatorOps) ... ok
test_compare_tensor_tensor (__main__.TestComparatorOps) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.476s

OK
```

Commit:
Update on "Quantized comparators"


Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Tests:

```
$> python test/test_quantized.py --v -- TestComparatorOps
test_compare_tensor_scalar (__main__.TestComparatorOps) ... ok
test_compare_tensor_tensor (__main__.TestComparatorOps) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.476s

OK
```

jamesr66a commented on aten/src/ATen/native/quantized/cpu/tensor_operators.cpp:
```diff
@@ -0,0 +1,54 @@
+                                Scalar other) { \
+  TORCH_CHECK(out.dtype() == at::ScalarType::Bool, \
+              "The 'out' tensor must have dtype 'torch.bool'"); \
+  const auto& self_dq = self.dequantize(); \
```
Why are all the results of dequantize() const references instead of values? Technically it's correct due to rvalue lifetime extension but it's confusing to the reader, especially in the presence of `auto`

zafartahirov replied:
I got used to having const ref whenever I use `auto` keyword -- this was an enforced rule in my previous life. Is it an anti-pattern ?

Commit:
Update on "Quantized comparators"


Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Tests:

```
$> python test/test_quantized.py --v -- TestComparatorOps
test_compare_tensor_scalar (__main__.TestComparatorOps) ... ok
test_compare_tensor_tensor (__main__.TestComparatorOps) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.476s

OK
```

Commit:
Update on "Quantized comparators"


Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Tests:

```
$> python test/test_quantized.py --v -- TestComparatorOps
test_compare_tensor_scalar (__main__.TestComparatorOps) ... ok
test_compare_tensor_tensor (__main__.TestComparatorOps) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.476s

OK
```

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/tensor_operators.cpp:
```diff
@@ -0,0 +1,54 @@
+  infer_size(self.sizes(), other.sizes()); \
+  TORCH_CHECK(out.dtype() == at::ScalarType::Bool, \
+              "The 'out' tensor must have dtype 'torch.bool'"); \
+  const Tensor self_dq = self.dequantize(); \
```
I don't think we need const since Tensor is just a pointer. maybe just use auto here?

dzhulgakov reviewed (commented):
As an inefficient implementation, looks ok. Can you add a TODO to make it efficient one? (might be a good task for someone new).

Also, how does it interact with type promotion like logic? Right now we can compare only quant-quant pair. It's fine to land it as a start, but it's good to figure out what the desired state should be. cc @gchanan @nairbv

gchanan commented:
We don't apply type promotion on comparison ops yet, because they are implemented in TH, but it's on @nairbv's list to address.

zafartahirov commented:
Type promotion in 

> As an inefficient implementation, looks ok. Can you add a TODO to make it efficient one? (might be a good task for someone new).
> 
> Also, how does it interact with type promotion like logic? Right now we can compare only quant-quant pair. It's fine to land it as a start, but it's good to figure out what the desired state should be. cc @gchanan @nairbv

Type promotion in quantized arithmetic is not well defined -- there is no way knowing what the output type should be (except very specific cases). For the comparators, it is easier. Especially with the inefficient solution like the one in this PR, it's just FP comparison.

Commit:
Update on "[quant] Quantized comparators"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24387

Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/tensor_operators.cpp:
```diff
@@ -0,0 +1,57 @@
+
+TODO: This is an inefficient implementation that uses `.dequantize`.
+      Need a more efficient implementation.
+*/
```
For scalar comparisons, can you quantize the scalar (much less expensive than dequantizing a tensor)?

zafartahirov replied:
A more efficient solution will be part of the upcoming PRs.

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -950,5 +950,73 @@ def test_qnnpack_linear(self, output_channels, X):
+    @given(A=hu.tensor(shapes=((3, 4, 5),),
+                       qparams=hu.qparams(dtypes=torch.quint8)),
+           B=hu.tensor(shapes=((5,), (1, 5), (1, 1, 5), (4, 5), (3, 4, 5)),
+                       qparams=hu.qparams(dtypes=torch.quint8)))
```
Can you sweep over dtypes to include qint8 and qint32? Does the comparison work for all scalar types (int/float/double?)

zafartahirov replied:
In the context of the current implementation it doesn't matter, because the implementation involves dequantization, any data type would work. In the upcoming PRs, I will add more efficient implementations, and appropriate randomized tests for them

raghuramank100 reviewed (changes requested):
Please see comments

Commit:
Update on "[quant] Quantized comparators"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24387

Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Commit:
Update on "[quant] Quantized comparators"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24387

Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)

Commit:
Update on "[quant] Quantized comparators"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24387

Differential Revision: [D16824421](https://our.internmc.facebook.com/intern/diff/D16824421)
